internal/db/db.go

```diff
@@ -104,7 +104,7 @@ func NewDB(
 	acp immutable.Option[acp.ACP],
 	lens client.LensRegistry,
 	options ...Option,
-) (client.DB, error) {
+) (*db, error) {
```

Comment:
There are a few reasons. One of them is because otherwise we have to make the client.DB API support everything even though some methods may just be useful internally. The maintenance overhead of that is large since we have to update the CLI and HTTP clients and test wrappers every time we update the interface. On top of that, there is no good reason to return an interface if the returned value is always of the same type. It's an unnecessary abstraction.

Reply:
> net.DB doesn't implement client.DB as GetNodeIdentityToken method is missing

This feels regressive, and to me it looks like we are forgetting the lessons learnt before `client.DB` was implemented by all packages - allowing the 3 clients to diverge (they now diverge in both directions, it is no longer a case of CLI contains embedded+some, CLI is now missing embedded stuff).

Returning a private type here, with public functions not declared in the `client` package also smashes the reasoning behind the `client` package in the first place - `GetNodeIdentityToken` is essentially a hidden public function that lives in an unexpected location, a location that we are very likely to forget about when making breaking changes (think semver).  And will unlikely be documented to the same standards as `client` members.

The multiple clients act as complexity multipliers.  The greater the divergence between clients, the greater that multiplier is.  This increases our maintenance costs (including documentation) and increases the learning required by users, in way that is far greater than the immediate X lines of code added.

IMO we need to be very very careful doing stuff like this, and seeing this so soon after another divergence (`purge` CLI command), makes it feel like we are falling back towards the time-sink that was our 3 un-interfaced, separate clients without making a conscious decision to do so.

Reply:
I actually did the change because I believe returning an interface where only one concrete type is ever returned is a needlessly restrictive abstraction.

The goal of an interface is to ensure that a given type implements a set of methods. **Not to limit** that type to the set of methods. There is a set of functionalities that we decided had to be covered by all clients/types (Go, CLI, HTTP) and that's where the interface is useful. If anything changes with those functionalities, it needs to change for all clients. 

There are some functionalities that will only ever make sense for one client. Purge for example only really makes sense for a server instance so it would be unnecessary overhead to require it for all clients through the client.DB interface.

What I think we can do is have a discussion on whether we want a given functionality to be part of the interface or not. But I really do think that some divergence between server and embedded is quite ok when it makes sense.

Reply:
> Not to limit that type to the set of methods. 

I quite strongly disagree with this, limiting a type to a sub-set of methods is a major reason for interface usage and is known as the interface segregation principle (the `I` in `SOLID`).

It reduces the surface area of the codebase and the number of pathways through it.  This reduces user-complexity, and the number of support tickets we will receive (as they wont be sending us tickets for funcs that are hidden by the interface).

Reply:
This is the `I` in `SOLID`
> A client should never be forced to implement an interface that it doesn’t use, or clients shouldn’t be forced to depend on methods they do not use.

In my interpretation, limiting a type to a sub-set of methods is not what the above means, at all.

Reply:
I see the signature of concrete type as an interface, by exposing the maximalist concrete interface you expose consuming code to problems in swapping it out for something else (essentially reducing the usefulness of defining the actual interface-interface).

That aside, IMO interface segregation goes beyond that one liner.  By segregating via interfaces you control the number pathways through the codebase and reduce complexity.

We don't want our users to be able to call all our internal funcs, and they probably don't want to see them either.

We end up with a public function that returns the public interface.  The user-code that consumes that type only ever directly depends-on/has-access-to the client/interface funcs, instead of a maximal concrete implementation.

The contract/interface of the public `NewDB` function then also remains identical to all other `NewDB` functions, allowing them to be used interchangeably without the swap being a breaking change.

Reply:
> I see the signature of concrete type as an interface, by exposing the maximalist concrete interface you expose consuming code to problems in swapping it out for something else (essentially reducing the usefulness of defining the actual interface-interface).

Why would `internal/DB` be swapped for something else? And if it is, what prevents it from supporting the `client.DB` interface?

> That aside, IMO interface segregation goes beyond that one liner. By segregating via interfaces you control the number pathways through the codebase and reduce complexity.

That one maybe. But I find that overly constraining to the point where one has to create more complex patterns to bypass the restrictions.

> We end up with a public function that returns the public interface. The user-code that consumes that type only ever directly depends-on/has-access-to the client/interface funcs, instead of a maximal concrete implementation.

We don't. `NewDB` is an internal function. Users don't have access to it.

> The contract/interface of the public NewDB function then also remains identical to all other NewDB functions, allowing them to be used interchangeably without the swap being a breaking change.

There is only one `NewDB` function. But even if we created many other DB alternatives, the important part is that those DB types cover the `client.DB` interface so that our http and cli clients along with the other internal types that need `client.DB` can use the `client.DB` methods. It doesn't matter that those other DB types provide more features.

Reply:
While I agree with the points raised by Andy, I don't think it's worth blocking the PR over. Specially in light of us doing the "moving fast experiment" I happy for this discussion to even resolve outside this PR. Just noting that the helper private function that was being called `newDB` was already using this signature before (irrelevant to the discussion I know).

Side question: I noticed that since the discussion thread was started the `*db` was made public (`*DB) now, curious for reasons behind that?

Reply:
> We don't. NewDB is an internal function. Users don't have access to it.

FFS lol - I forgot we moved it, sorry. 90% of me bothering about this is because I thought it was still a public func and you were exposing internals to users.

> There is only one NewDB function

There are ones for the other 2 clients, and one from the node package no? They may have different names atm, but they are all basically the same func (and should probs have the same name).

> Why would internal/DB be swapped for something else?

The callers to the `NewDB` func(s) may wish to swap them out - this becomes much harder if the type changes and extra/non-interface stuff on the return value is used.

But anyway, this is just internal stuff, the extra exposure does make stuff easier, and we (in theory :)) should really be able to tell what the newly exposed stuff does anyway.

Happy for this convo to be resolved lol

Reply:
Just adding for historical sake since I think Andy has dropped this (and as shahzad pointed out, we're aiming not to get bogged down during this cycle) - I think adding compile time checks to the `db` package that make sure the `DB` type implements the `client.DB` is all that is really necessary. That along with the downstream callers (internal to the repo) having the option to assign/cast to `client.DB` for their own safety if they want should cover us for most of our cases.

Fred and I briefly discussed this whole return type a few weeks ago and one thing it raised is we currently don't have a proper "embedded" client flow ever since we moved the `db` package to `internal`. The way the integration tests access the embedded flow is to instanciate a `node.Node` instance, then getting the `client.DB` instance from it. But we should have a dedicated and public flow for instantiating the `client.DB` without the complexitity of the `node.New` constructor as follow up work to this